libcontainer/nsenter/nsexec.c

```diff
@@ -739,54 +739,65 @@ void send_fd(int sockfd, int fd)
 		bail("failed to send fd %d via unix socket %d", fd, sockfd);
 }
 
-void receive_mountsources(int sockfd, char *mountsources, size_t mountsources_len)
+void receive_mountsources(int sockfd)
 {
-	char *mount_file_fds_str;
+	char *mount_fds_str;
 	char *new_fd_str;
 	char *saveptr = NULL;
 
-	mount_file_fds_str = getenv("_LIBCONTAINER_MOUNT_FDS");
+	mount_fds_str = getenv("_LIBCONTAINER_MOUNT_FDS");
 
 	// container_linux.go shouldSendMountSources() decides if mount sources
 	// should be pre-opened (O_PATH) and passed via SCM_RIGHTS
-	if (mount_file_fds_str == NULL || *mount_file_fds_str == '\0')
-		return;
-	if (mountsources == NULL)
+	if (mount_fds_str == NULL || *mount_fds_str == '\0')
 		return;
 
 	// make a copy because strtok_r modifies the variable
-	mount_file_fds_str = strdup(mount_file_fds_str);
-	if (mount_file_fds_str == NULL) {
-		bail("Can't allocate memory to strdup mount_file_fds_str");
+	mount_fds_str = strdup(mount_fds_str);
+	if (mount_fds_str == NULL) {
+		bail("Can't allocate memory to strdup mount_fds_str");
 	}
+	// Save the pointer, as we will modify this and need to free() it later.
+	char *orig_mount_fds_str = mount_fds_str;
 
-	new_fd_str = strtok_r(mount_file_fds_str, ";", &saveptr);
+	// This is a json encoded list, it should not be shorter than: []
+	if (strlen(mount_fds_str) < 2) {
+		bail("Malformed _LIBCONTAINER_MOUNT_FDS: Should have at least 2 characters, got: %s", mount_fds_str);
+	}
+	// Remove the starting '['
+	if (mount_fds_str[0] != '[') {
+		bail("Malformed _LIBCONTAINER_MOUNT_FDS: Missing starting '['");
+	}
+	mount_fds_str++;
 
-	char *mountsources_end = mountsources + mountsources_len;
-	while (mountsources < mountsources_end) {
-		int new_fd;
+	// Remove the ending ']'
+	if (mount_fds_str[strlen(mount_fds_str) - 1] != ']') {
+		bail("Malformed _LIBCONTAINER_MOUNT_FDS: Missing ending ']'");
+	}
+	mount_fds_str[strlen(mount_fds_str) - 1] = '\0';
 
-		// $_LIBCONTAINER_MOUNT_FDS might contain empty entries in
-		// the ";"-separated list
-		while (new_fd_str != NULL && new_fd_str[0] == '\0') {
-			new_fd_str = strtok_r(NULL, ";", &saveptr);
-		}
-		if (new_fd_str == NULL)
-			break;
+	new_fd_str = strtok_r(mount_fds_str, ",", &saveptr);
+	while (new_fd_str != NULL) {
+		int new_fd = atoi(new_fd_str);
 
-		if (mountsources[0] == '\0') {
-			mountsources++;
+		// _LIBCONTAINER_MOUNT_FDS contains entries that are -1 (invalid
+		// fds), we need to ignore them.
+		if (new_fd == -1) {
+			new_fd_str = strtok_r(NULL, ",", &saveptr);
 			continue;
 		}
+		// XXX: atoi() returns 0 on error and, by how fds are allocated
+		// in runc, this can never be 0. Bail if that ever happens.
```

Comment:
I'd rewrite it using `strtol`, as we can detect errors, plus we avoid using `strtok` and `strdup`.

Something like this (this example just prints all the numbers decoded):
```c
mount_fds_str = getenv("_LIBCONTAINER_MOUNT_FDS");
void decode_fds(char *str) {
        if (*str++ != '[') {
                printf("malformed json, missing [\n");
                return;
        }

        while (1) {
                int n;
                char *endp;

                n = strtol(str, &endp, 10);
                if (str == endp) {
                        printf("malformed json, not a number\n");
                        return;
                }
                printf("%d", n);
                if (*endp == ']') {
                        break;
                }
                if (*endp == '\0') {
                        printf("malformed json, missing ]\n");
                        return;
                }
                printf(",");
                str = endp+1; // skip ,
        }
        printf("\n");
}
```

Reply:
Nice idea. Will push a version using this as inspiration, thanks!

Reply:
Fixed, it actually looks quite nice now. Thanks!